File: runtime/rs/src/option.rs
// option extern — Rust native implementations

pub fn almide_rt_option_map<T: Clone, U>(o: Option<T>, f: impl Fn(T) -> U) -> Option<U> { o.map(f) }
pub fn almide_rt_option_and_then<T: Clone, U>(o: Option<T>, f: impl Fn(T) -> Option<U>) -> Option<U> { o.and_then(f) }
pub fn almide_rt_option_unwrap_or<T>(o: Option<T>, default: T) -> T { o.unwrap_or(default) }
pub fn almide_rt_option_is_some<T>(o: &Option<T>) -> bool { o.is_some() }
pub fn almide_rt_option_is_none<T>(o: &Option<T>) -> bool { o.is_none() }
pub fn almide_rt_option_flatten<T>(o: Option<Option<T>>) -> Option<T> { o.flatten() }
pub fn almide_rt_option_to_result<T>(o: Option<T>, err: String) -> Result<T, String> { o.ok_or(err) }
pub fn almide_rt_option_unwrap_or_else<T>(o: Option<T>, f: impl Fn() -> T) -> T { o.unwrap_or_else(f) }
pub fn almide_rt_option_filter<T: Clone>(o: Option<T>, f: impl Fn(T) -> bool) -> Option<T> { o.filter(|x| f(x.clone())) }
pub fn almide_rt_option_zip<T, U>(a: Option<T>, b: Option<U>) -> Option<(T, U)> { a.zip(b) }
pub fn almide_rt_option_or_else<T>(o: Option<T>, f: impl Fn() -> Option<T>) -> Option<T> { o.or_else(f) }
pub fn almide_rt_option_to_list<T>(o: Option<T>) -> Vec<T> {
    match o { Some(v) => vec![v], None => vec![] }
}

File: runtime/ts/option.ts
// Option module — TS runtime
// Option[T] = T | null (some(x) = x, none = null)

const __almd_option = {
  map<T, U>(o: T | null, f: (x: T) => U): U | null {
    return o !== null ? f(o) : null;
  },
  flat_map<T, U>(o: T | null, f: (x: T) => U | null): U | null {
    return o !== null ? f(o) : null;
  },
  flatten<T>(o: (T | null) | null): T | null {
    return o !== null ? o : null;
  },
  unwrap_or<T>(o: T | null, def: T): T {
    return o !== null ? o : def;
  },
  unwrap_or_else<T>(o: T | null, f: () => T): T {
    return o !== null ? o : f();
  },
  is_some<T>(o: T | null): boolean {
    return o !== null;
  },
  is_none<T>(o: T | null): boolean {
    return o === null;
  },
  to_result<T>(o: T | null, err: string): T {
    if (o !== null) return o;
    throw new Error(err);
  },
  filter<T>(o: T | null, f: (x: T) => boolean): T | null {
    return o !== null && f(o) ? o : null;
  },
  zip<T, U>(a: T | null, b: U | null): [T, U] | null {
    return a !== null && b !== null ? [a, b] : null;
  },
  or_else<T>(o: T | null, f: () => T | null): T | null {
    return o !== null ? o : f();
  },
  to_list<T>(o: T | null): T[] {
    return o !== null ? [o] : [];
  },
};
